python/dgl/data/adapter.py

```diff
@@ -108,3 +114,181 @@ def __getitem__(self, idx):
 
     def __len__(self):
         return 1
+
+
+
+def negative_sample(g, num_samples):
+    """Random sample negative edges from graph, excluding self-loops"""
+    num_nodes = g.num_nodes()
+    redundancy = _calc_redundancy(
+        num_samples, g.num_edges(), num_nodes ** 2)
+    sample_size = int(num_samples*(1+redundancy))
+    edges = np.random.randint(0, num_nodes, size=(2, sample_size))
+    edges = np.unique(edges, axis=1)
+    # remove self loop
+    mask_self_loop = edges[0] == edges[1]
+    # remove existing edges
+    has_edges = F.asnumpy(g.has_edges_between(edges[0], edges[1]))
+    mask = ~(np.logical_or(mask_self_loop, has_edges))
+    edges = edges[:, mask]
+    if edges.shape[1] >= num_samples:
+        edges = edges[:, :num_samples]
+    return edges
+
+
+class AsEdgePredDataset(DGLDataset):
+    """Repurpose a dataset for edge prediction task.
+
+    The created dataset will include data needed for link prediction. 
+    It will keep only the first graph in the provided dataset and
+    generate train/val/test edges according to the given split ratio,
+    and the correspondent negative edges based on the neg_ratio. The generated
+    edges will be cached to disk for fast re-loading. If the provided split ratio
+    differs from the cached one, it will re-process the dataset properly.
+
+    Parameters
+    ----------
+    dataset : DGLDataset
+        The dataset to be converted.
+    split_ratio : (float, float, float), optional
+        Split ratios for training, validation and test sets. Must sum to one.
+    neg_ratio : int, optional
+        Indicate how much negative samples to be sampled
+        The number of the negative samples will be neg_ratio * num_positive_edges.
+
+    Returns
+    -------
+    DGLDataset
+        A new dataset with only one graph. Train/val/test masks are stored in the
+        ndata of the graph.
+
+    Examples
+    --------
+    >>> ds = dgl.data.CoraGraphDataset()
+    >>> print(ds)
+    Dataset("cora_v2", num_graphs=1, save_path=...)
+    >>> new_ds = dgl.data.AsNodePredDataset(ds, [0.8, 0.1, 0.1])
+    >>> print(new_ds)
+    Dataset("cora_v2-as-edgepred", num_graphs=1, save_path=/home/ubuntu/.dgl/cora_v2-as-edgepred)
+    >>> print(hasattr(new_ds, "get_test_edges"))
+    True
+    """
+
+    def __init__(self,
+                 dataset,
+                 split_ratio=[0.8, 0.1, 0.1],
+                 neg_ratio=3,
+                 add_self_loop=True,
+                 **kwargs):
+        self.g = dataset[0]
+        self.dataset = dataset
+        self.split_ratio = split_ratio
+        self.neg_ratio = neg_ratio
+        self.add_self_loop = add_self_loop
+        super().__init__(dataset.name + '-as-edgepred', hash_key=(neg_ratio, split_ratio, add_self_loop), **kwargs)
```

Comment:
If `hash_key` is designed for proper re-loading. Please help change `AsNodePred` as well. Make sure you've tested them properly.